server/src/main/java/io/crate/execution/ddl/tables/TransportAddColumnAction.java

```diff
@@ -0,0 +1,315 @@
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.IndexTemplateMetadata;
+import org.elasticsearch.cluster.metadata.MappingMetadata;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.inject.Singleton;
+import org.elasticsearch.common.settings.IndexScopedSettings;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.index.mapper.DocumentMapper;
+import org.elasticsearch.index.mapper.MapperService;
+import org.elasticsearch.indices.IndicesService;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+import static io.crate.metadata.cluster.AlterTableClusterStateExecutor.resolveIndices;
+import static org.elasticsearch.index.mapper.TypeParsers.DOC_VALUES;
+
+@Singleton
+public class TransportAddColumnAction extends AbstractDDLTransportAction<AddColumnRequest, AcknowledgedResponse> {
+
+    private static final String ACTION_NAME = "internal:crate:sql/table/add_column";
+    private final NodeContext nodeContext;
+    private final IndicesService indicesService;
+
+    @Inject
+    public TransportAddColumnAction(TransportService transportService,
+                                    ClusterService clusterService,
+                                    IndicesService indicesService,
+                                    ThreadPool threadPool,
+                                    NodeContext nodeContext) {
+        super(ACTION_NAME,
+            transportService,
+            clusterService,
+            threadPool,
+            AddColumnRequest::new,
+            AcknowledgedResponse::new,
+            AcknowledgedResponse::new,
+            "add-column");
+        this.nodeContext = nodeContext;
+        this.indicesService = indicesService;
+    }
+
+    /**
+     * Class for gathering mapping data in metadata compatible format.
+     * AddColumnRequest has multiple columns and each of them can be a root of an object column tree and thus
+     * can have many not-null/primary key/generated constraints.
+     * We gather everything here in order to do only one column-tree traversal and call mergeInto only once for each type of constraint.
+     */
+    private final class ConvertedRequest {
+
+        private final ArrayList<String> primaryKeys = new ArrayList<>();
+        private final ArrayList<String> notNullColumns = new ArrayList<>();
+        private final LinkedHashMap<String, String> generatedColumns = new LinkedHashMap<>();
+        private final HashMap<String, Object> topColumns = new HashMap<>();
```

Comment:
Any chance to build the mapping directly from the request to reduce the overhead a bit?

Reply:
map itself is being created from the request (we traverse with `StreamableColumInfo`) but without collecting each constraint we would need to merge constraint right away in each traversing iteration - meaning that we will do 
```
  Maps.mergeInto(existingMapping, "_meta", List.of("primary_keys")...
  Maps.mergeInto(existingMapping, "_meta", List.of("not_null")...
  Maps.mergeInto(existingMapping, "_meta", List.of("checks")...
```
`tree_size` times. in case of dynamic mapping update it can be multiple columns, each can be some big object - now we add all at once at the end. idk, both ways have flaws.

Reply:
It would be possible to do the mapping merge more concretely and write to the `_meta.primary_keys` directly. For example something like this:

```
        Map<String, Object> meta = (Map<String, Object>) existingMapping.get("_meta");
        List<String> primaryKeys = (List<String>) meta.get("primary_keys");

        for (var column : request.columns()) {
            if (column.isPrimaryKey()) {
                if (primaryKeys == null) {
                    primaryKeys = new ArrayList<>();
                    meta.put("primary_keys", primaryKeys);
                }
                primaryKeys.add(...)
            }
        }
```

That would avoid all that overhead

Reply:
done in https://github.com/crate/crate/pull/12806/commits/a835d593b20ecf58b3e21b16dcbdc5a56638f372

please ignore Itest, added to debug some issues, will get rid of such tests later if replaceable by unit tests.

I recalled why I gathered all PK/not nulls  - it was not really `mergeInto` being my concern (logic is very similar to explicit merge from ^) but rather saving of several traversals - we cannot iterate over top level columns,  we have to traverse deep down and make it twice if table is partitioned. I I still have properties map in some variable as it doesn't make sense to compute it twice (partioned table and then regular `updateMapping`)

dynamic mapping update cannot have constraints and add column can have max 1 of each constraint so I got rid of map/list with 1 element but now traverse thrice (1 time build properties map + 2 times constraints). But what is better is that we don't have extra entity ConvertedRequest - but I think saving traversals is more important as we can have more columns than constraints.